### PR TITLE
Evo2 prediction and inference scripts, and example notebooks.

### PR DESCRIPTION
**Pull request by jstjohn**
### Description

This PR adds back prediction and inference scripts, as well as the fine-tuning and brca zero shot fine-tuning notebooks. This also adds in support for batched vortex generation.

#### Usage

See `examples/*.ipynb` in the `evo2_megatron` recipe for usage.

### Type of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels. By default, only basic unit tests are run.

- [ciflow:skip](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:skip) - Skip all CI tests for this PR
- [ciflow:notebooks](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:notebooks) - Run Jupyter notebooks execution tests for bionemo2
- [ciflow:slow](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:slow) - Run slow single GPU integration tests marked as @pytest.mark.slow for bionemo2
- [ciflow:all](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all) - Run all tests (unit tests, slow tests, and notebooks) for bionemo2. This label can be used to enforce running tests for all bionemo2.
- [ciflow:all-recipes](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all-recipes) - Run tests for all recipes (under bionemo-recipes). This label can be used to enforce running tests for all recipes.

Unit tests marked as `@pytest.mark.multi_gpu` or `@pytest.mark.distributed` are not run in the PR pipeline.

For more details, see [CONTRIBUTING](CONTRIBUTING.md)

> [!NOTE]
> By default, only basic unit tests are run. Add appropriate labels to enable an additional test coverage.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Pre-submit Checklist

- [ ] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [ ] I have added/updated tests as needed
- [ ] All existing tests pass successfully

## Summary by CodeRabbit

* **New Features**
  * Full inference engine and CLI for text generation, prediction workflow, and a simple token-generation helper.
  * New training flag to disable FP8 weight gradients.

* **Bug Fixes**
  * Corrected multi-batch decoding behavior.
  * Ensured final normalization is always defined and preprocessing/postprocessing flags honor explicit disables.

* **Improvements**
  * Tokenizer decoding switched to a Fuse decoder.
  * Enhanced GPU cleanup and robust test utilities; expanded, runnable test suites.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by jstjohn** [comment]
@coderabbitai review

**Comment by jstjohn** [comment]
@coderabbitai review

File: bionemo-recipes/recipes/evo2_megatron/tests/bionemo/evo2/test_stop_and_go.py [hunk 1]
```diff
@@ -240,169 +194,3 @@ def test_stop_and_go(
     assert first_loss_run2 - last_loss_run1 < 0.1, (
         f"Run 2 first loss {first_loss_run2} is not better than run 1 last loss {last_loss_run1} by no worse than 0.1"
     )
-
-
-@pytest.mark.slow
-def test_fine_tuning(
```
Comment (jstjohn) [inline bionemo-recipes/recipes/evo2_megatron/tests/bionemo/evo2/test_stop_and_go.py:246]: Moved to `test_train.py` FYI.

File: bionemo-recipes/recipes/evo2_megatron/tokenizers/nucleotide_fast_tokenizer_256/tokenizer.json [hunk 2]
```diff
@@ -136,7 +136,9 @@
       }
     }
   },
-  "decoder": null,
+  "decoder": {
```
Comment (jstjohn) [inline bionemo-recipes/recipes/evo2_megatron/tokenizers/nucleotide_fast_tokenizer_256/tokenizer.json:139]: FYI this fixes the issue where when you detokenize by default the tokenizer would insert spaces between the "words".

File: bionemo-recipes/recipes/evo2_megatron/src/bionemo/evo2/models/evo2_provider.py [hunk 3]
```diff
@@ -511,8 +511,20 @@ def provide(self, pre_process=None, post_process=None, vp_stage=None) -> MCoreHy
...
+            pre_process=(
+                False
+                if self.pre_process is False
+                else (pre_process if pre_process is not None else parallel_state.is_pipeline_first_stage())
+            ),
+            post_process=(
+                False
+                if self.post_process is False
+                else (post_process if post_process is not None else parallel_state.is_pipeline_last_stage())
+            ),
```
Comment (cspades) [inline bionemo-recipes/recipes/evo2_megatron/src/bionemo/evo2/models/evo2_provider.py:527]: Not sure if relevant, but this changes the original OR logic:
```
pre_process or parallel_state.is_pipeline_first_stage()
```
Now, you can set `pre_process` to `False` even if it's the first pipeline stage, and same for `post_process`. Intentional?

**Review by cspades — APPROVED** [review]
One observation that looks weird, approving pending that confirm.

File: bionemo-recipes/recipes/evo2_megatron/src/bionemo/evo2/models/evo2_provider.py (hunk 3 above)
Comment (jstjohn) [reply bionemo-recipes/recipes/evo2_megatron/src/bionemo/evo2/models/evo2_provider.py:527]: Yes that's intentional. Post process has the use case of gathering embeddings.
